mlir/lib/Dialect/MIOpen/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -112,42 +112,80 @@ struct BlockwiseGemmRewritePattern
+    int64_t nCuwavesPerBlock = op.nCuwavesPerBlockAttr().getInt();
+    int64_t numCuwaves = mCuwavesPerBlock * nCuwavesPerBlock;
+    int64_t blockSize = numCuwaves * cuwaveLen;
+
     int64_t mRepeat = mC / mPerThread;
     int64_t nRepeat = nC / nPerThread;
 
-    LLVM_DEBUG(llvm::dbgs() << "M: " << mC << "\n"
-                            << "NRepeat: " << mRepeat << "\n"
-                            << "MPerThread: " << mPerThread << "\n"
-                            << "N: " << nC << "\n"
-                            << "NRepeat: " << nRepeat << "\n"
-                            << "NPerThread: " << nPerThread << "\n");
-
-    TopDownTMBuilder strideLDSBufferA(b,
-                                      {"k", "mRepeat", "mPerThread", "kpack"},
-                                      {k, mRepeat, m / mRepeat, kPack}, loc);
-    strideLDSBufferA.passThrough("k");
-    strideLDSBufferA.embed("m", 1, m, {"mRepeat", "mPerThread"},
-                           {mRepeatStride, 1});
-    strideLDSBufferA.passThrough({"kpack"}, {2}, {"kpack"});
-    TransformMapAttr strideLDSBufferAAttr = strideLDSBufferA.get();
-
-    TopDownTMBuilder strideLDSBufferB(b,
-                                      {"k", "nRepeat", "nPerThread", "kpack"},
-                                      {k, nRepeat, n / nRepeat, kPack}, loc);
-    strideLDSBufferB.passThrough("k");
-    strideLDSBufferB.embed("n", 1, n, {"nRepeat", "nPerThread"},
-                           {nRepeatStride, 1});
-    strideLDSBufferB.passThrough({"kpack"}, {2}, {"kpack"});
-    TransformMapAttr strideLDSBufferBAttr = strideLDSBufferB.get();
+    LLVM_DEBUG(llvm::dbgs()
+               << "M: " << m << "\n"
+               << "mRepeat: " << mRepeat << "\n"
+               << "mCuwavesPerBlock: " << mCuwavesPerBlock << "\n"
+               << "mThreadsPerCuwave: " << mThreadsPerCuwave << "\n"
+               << "mPerThread: " << mPerThread << "\n"
+               << "n: " << n << "\n"
+               << "nRepeat: " << nRepeat << "\n"
+               << "nCuwavesPerBlock: " << nCuwavesPerBlock << "\n"
+               << "nThreadsPerCuwave: " << nThreadsPerCuwave << "\n"
+               << "nPerThread: " << nPerThread << "\n");
+
+    auto ldsTidSplitter = [&](StringRef repeatName, int64_t repeatLen,
+                              StringRef perThreadName,
+                              int64_t perThreadLen) -> TopDownTMBuilder {
+      TopDownTMBuilder splitTidForLDS(
+          b, {"k", repeatName, "tid", perThreadName, "kpack"},
+          {k, repeatLen, blockSize, perThreadLen, kPack}, loc);
+      splitTidForLDS.passThrough({"k", repeatName});
+      splitTidForLDS.merge({"m_cuwaves", "n_cuwaves", "m_cuwave", "n_cuwave"},
+                           {2, 3, 4, 5}, "tid",
+                           {mCuwavesPerBlock, nCuwavesPerBlock,
+                            mThreadsPerCuwave, nThreadsPerCuwave},
+                           /*isUnfold=*/true);
```

Comment:
[Question] Are you trying to construct fewer builder by incorporating A and B's dimensions together and then later ignore `m` or `n`?

Reply:
The way this math works is you split the thread ID into which CU wave (group of 16) you're on and which thread you're in within that wave. Then, both the larger waves and the threads within them form a 2D grid.

That is, which `m` you have is `tid / (cuwaveLen * nCuwavesPerBlock) + (tid % cuwaveLen) / nCuwavesPerBlock` , while the `n` is `(tid / cuwaveLen) % nCuwavesPerBlock + (tid % nThreadsPerCuwave)`

So we need to split the TID up into the 4-way grid (m_cuwaves x n_cuwaves x m_cuwave x n_cuwave) and then grab the relevant pair of numbers.